lsp/nls/src/requests/completion.rs

```diff
@@ -1,16 +1,198 @@
+fn find_record_contract(
+    linearization: &Vec<LinearizationItem<Types>>,
+    id: usize,
+) -> Option<Contract> {
+    linearization.iter().find_map(|item| match &item.meta {
+        Some(MetaValue { contracts, .. }) if item.id == id => {
+            contracts.iter().find_map(|contract| {
+                if let Types(AbsType::Record(..)) = contract.types {
+                    Some(contract.clone())
+                } else {
+                    None
+                }
+            })
+        }
+        _ if item.id == id => match item.kind {
+            TermKind::Declaration(_, _, ValueState::Known(new_id))
+            | TermKind::Usage(UsageState::Resolved(new_id)) => {
+                find_record_contract(&linearization, new_id)
+            }
+            _ => None,
+        },
+        _ => None,
+    })
+}
+
+/// Extract identifiers from a row type which forms a record.
+/// Ensure that ty is really a row type.
+fn extract_ident(ty: &Box<Types>) -> Vec<Ident> {
+    ty.iter_as_rows()
+        .filter_map(|item| match item {
+            RowIteratorItem::Row(ident, _) => Some(ident.clone()),
+            RowIteratorItem::Tail(..) => None,
+        })
+        .collect::<Vec<_>>()
+}
+
+/// Get the identifier before the `.`, for record completion.
+fn get_identifier(text: &str) -> Result<String, ResponseError> {
+    lazy_static! {
+        // unwrap is safe here because we know this is a correct regex.
+        // This regexp must be the reverse of the regexp in the lexer (nickel_lang::parser::lexer)
+        // to correctly parse identifiers.
+        static ref RE: regex::Regex = regex::Regex::new(r"[_a-zA-Z0-9-']*[a-zA-Z]_?").unwrap();
+    }
+
+    let text: String = text
+        .chars()
+        .rev()
+        .skip_while(|c| *c == '.') // skip . (if any)
+        .collect();
+
+    let name = RE.find(&text).ok_or(ResponseError {
+        code: ErrorCode::InternalError as i32,
+        message: "Couldn't get identifier for record completion".to_owned(),
+        data: None,
+    })?;
+
+    Ok(name.as_str().chars().rev().collect())
+}
+
+fn remove_duplicates(items: &Vec<CompletionItem>) -> Vec<CompletionItem> {
+    let mut seen: Vec<CompletionItem> = Vec::new();
+    for item in items {
+        if !seen.iter().any(|seen_item| seen_item.label == item.label) {
+            seen.push(item.clone())
+        }
+    }
+    seen
+}
+
+fn get_completion_identifiers(
+    source: &str,
+    trigger: Option<&str>,
+    linearization: &Completed,
+    item: &LinearizationItem<Types>,
+) -> Result<Vec<CompletionItem>, ResponseError> {
+    let lin_env = &linearization.lin_env;
+    let in_scope = match trigger {
+        // Record Completion
+        Some(server::DOT_COMPL_TRIGGER) => {
+            let name = get_identifier(source)?;
+            let name_id = lin_env
+                .iter()
+                .find(|(ident, _)| ident.label() == name)
+                .map(|(_, id)| *id);
```

Comment:
Why do we still do an `iter()` here, which is linear-time, instead of using `Environment::get` ?

Reply:
Oh, because the `lin_env` is an env of `Ident`. Then maybe it should be an environment of type  `Environment<String, usize>`? Otherwise it feels a bit useless to use an `environment` but to iterate over it as a vector.

Reply:
A second solution is to create an Ident from the string. Thanks to string interning both approaches should be close, so pick whichever is simpler from a code point of view.

Reply:
Nice, thank you! Fixed.